src/Toolchain.re

```diff
@@ -0,0 +1,759 @@
+  };
+  type stdout = string;
+  let binPath = c => c.cmd;
+  let make = (~env, ~cmd) => {
+    switch (Js.Dict.get(env, "PATH")) {
+    | None => Error(pathMissingFromEnv) |> Js.Promise.resolve
+    | Some(path) =>
+      let cmds = Sys.unix ? [|cmd|]: [|cmd ++ ".exe", cmd ++ ".cmd"|];
+      cmds
+      |> Array.map(~f=cmd => {
+        String.split(~on=env_sep, path)
+        |> Array.fromList
+        |> Js.Array.map(p => Filename.concat(p, cmd))
+      })
+      |> Array.concatenate
+      |> Js.Array.map(p => { 
+          Fs.exists(p)
+          |> Js.Promise.then_(exists => Js.Promise.resolve((p, exists)))
+        })
+      |> Js.Promise.all
+      |> Js.Promise.then_(Js.Promise.resolve << Js.Array.filter(((p, exists)) => exists))
+      |> Js.Promise.then_(
+           Js.Promise.resolve
+           << (
+             fun 
+             | [] => Error({j| Command "$cmd" not found |j})
+             | [(cmd, exists), ..._rest] => Ok({cmd, env})
+           ) << Array.toList,
+         )
+    };
+  };
+  let output = (~args, ~cwd, {cmd, env}) => {
+    let shellString =
+      Js.Array.concat(args, [|cmd|]) |> Js.Array.joinWith(" ");
+    Js.log(shellString);
+    ChildProcess.exec(shellString, ChildProcess.Options.make(~cwd, ~env, ()))
+    |> Js.Promise.then_(
+         Js.Promise.resolve
+         << (
+           fun
+           | Error(e) => e |> ChildProcess.E.toString |> R.fail
+           | Ok((exitCode, stdout, stderr)) =>
+             if (exitCode == 0) {
+               Ok(stdout);
+             } else {
+               Error(
+                 {j| Command $cmd failed:
+exitCode: $exitCode
+stderr: $stderr
+|j},
+               );
+             }
+         ),
+       );
+  };
+};
+
+/* Package managers that can install native Reason/
+   OCaml packages for us */
+module PackageManager: {
+  type t;
+  type a = t;
+  type spec;
+  module type T = {
+    let name: string;
+    let lockFile: Fpath.t;
+    let make:
+      (
+        ~env: Js.Dict.t(string),
+        ~root: Fpath.t,
+        ~discoveredManifestPath: Fpath.t
+      ) =>
+      Js.Promise.t(result(spec, string));
+  };
+  module Esy: T;
+  module Opam: T;
+  let ofName: string => result(t, string);
+  let toName: t => string;
+  let specOfName:
+    (
+      ~env: Js.Dict.t(string),
+      ~name: string,
+      ~root: Fpath.t,
+      ~discoveredManifestPath: Fpath.t
+    ) =>
+    Js.Promise.t(result(spec, string));
+  let make:
+    (
+      ~env: Js.Dict.t(string),
+      ~root: Fpath.t,
+      ~discoveredManifestPath: Fpath.t,
+      ~t: t
+    ) =>
+    Js.Promise.t(result(spec, string));
+  let setupToolChain: spec => Js.Promise.t(result(unit, string));
+
+  module PackageManagerSpecTuple: {
+    type t = (a, Utils.Fpath.t);
+    let compare: (('a, 'b), ('a, 'b)) => int;
+  };
+  module PackageManagerSpecTupleSet: {
+    type elt = PackageManagerSpecTuple.t;
+    type t = Set.Make(PackageManagerSpecTuple).t;
+    let empty: t;
+    let is_empty: t => bool;
```

Comment:
Why do we list out the entire signature of Set? Can’t we use an include?

Reply:
Sure!